common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.cpp

```diff
@@ -509,9 +552,146 @@ bool base_wlan_hal_dwpal::attach_ctrl_interface(int vap_id)
 
 bool base_wlan_hal_dwpal::process_nl_events()
 {
-    LOG(DEBUG) << __func__ << "isn't implemented by derived and shouldn't be called";
-    return false;
+    // process all NL events
+    if (!m_dwpal_nl_ctx) {
+        LOG(ERROR) << "Invalid WPA Control socket used for nl events (m_dwpal_ctx == nullptr)";
+        return false;
+    }
+
+    // Passing a lambda with capture is not supported for standard C function
+    // pointers. As a workaround, we create a static (but thread local) wrapper
+    // function that calls the capturing lambda function.
+    static __thread std::function<DWPAL_Ret(struct nl_msg * msg)> nl_handler_cb_wrapper;
+    nl_handler_cb_wrapper = [&](struct nl_msg *msg) -> DWPAL_Ret {
+        if (!process_dwpal_nl_event(msg)) {
+            LOG(ERROR) << "User's netlink handler function failed!";
+            return DWPAL_FAILURE;
+        }
+        return DWPAL_SUCCESS;
+    };
+    auto nl_handler_cb = [](struct nl_msg *msg) -> DWPAL_Ret { return nl_handler_cb_wrapper(msg); };
+
+    if (m_fd_nl_events > 0) {
+        //parsing will be done in callback function
+        if (dwpal_driver_nl_msg_get(m_dwpal_nl_ctx, DWPAL_NL_UNSOLICITED_EVENT, NULL,
+                                    nl_handler_cb) == DWPAL_FAILURE) {
+            LOG(ERROR) << " dwpal_driver_nl_msg_get failed,"
+                       << " ctx=" << m_dwpal_nl_ctx;
+            detach();
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool base_wlan_hal_dwpal::dwpal_nl_cmd_set(const std::string &ifname, unsigned int nl_cmd,
+                                           unsigned char *vendor_data, size_t vendor_data_size)
+{
+    if ((vendor_data_size > 0) && (vendor_data == NULL)) {
+        LOG(ERROR) << __FUNCTION__ << "ifname or vendor_data is NULL ==> Abort!\n";
+        return false;
+    }
+
+    if (dwpal_driver_nl_cmd_send(m_dwpal_nl_ctx, DWPAL_NL_UNSOLICITED_EVENT, (char *)ifname.c_str(),
+                                 NL80211_CMD_VENDOR, DWPAL_NETDEV_ID,
+                                 (enum ltq_nl80211_vendor_subcmds)nl_cmd, vendor_data,
+                                 vendor_data_size) != DWPAL_SUCCESS) {
+        LOG(ERROR) << __FUNCTION__ << "ERROR for cmd = " << nl_cmd;
+        return false;
+    }
+
+    return true;
+}
+
+bool base_wlan_hal_dwpal::dwpal_nl_cmd_get(const std::string &ifname, unsigned int nl_cmd,
+                                           unsigned char *out_buffer, size_t *out_buffer_size,
+                                           const size_t max_buffer_size)
+{
+
+    if ((out_buffer_size == NULL) || (out_buffer == NULL)) {
```

Comment:
`nullptr`

Reply:
This API should be simplified to use a single size_t inout length parameter:
```
bool base_wlan_hal_dwpal::dwpal_nl_cmd_get(const std::string &ifname, unsigned int nl_cmd,
                                           unsigned char *buf, size_t &buf_len)
```
so here buf_len is set by the caller to the actual size of the buffer, which is the maximum size of the buffer, and the function updates the value of `buf_len` to the actual size of the returned reply.

Reply:
I agree that its not pretty code but prefer not to use i/o parameter.
Changed according to  @vitalybu [suggestion ](https://github.com/prplfoundation/prplMesh/pull/637/commits/5b40d855d32bee57d44915d76659316d8a860933#r364738317)